### PR TITLE
Implement generic signature reconstruction for methods and use it in RestClient Reactive

### PR DESCRIPTION
**Pull request by Ladicek**
In addition, update Gizmo to the latest version that has complete
support for generating annotations.

### CODE REVIEW

**Comment by Ladicek** [comment]
Draft to let CI run in my fork first, but otherwise complete.

For review, I added some people that I know have been involved in bytecode generation. If you can think of better / additional reviewers, that would be great!

File: core/deployment/src/main/java/io/quarkus/deployment/util/JandexUtil.java [hunk 1]
```diff
@@ -375,4 +380,219 @@ public ClassNotIndexedException(DotName dotName) {
...
+            result.append(':');
+        }
+        for (Type bound : typeParameter.bounds()) {
+            result.append(':');
+            typeSignature(bound, result);
+        }
+    }
+
+    private static boolean hasImplicitObjectBound(TypeVariable typeParameter) {
+        // TODO is there a better way? :-/
```
Comment (Ladicek) [inline core/deployment/src/main/java/io/quarkus/deployment/util/JandexUtil.java:494]: @n1hility any help here? This is the only thing I wasn't able to reconstruct from Jandex's public API.

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java [hunk 2]
```diff
@@ -307,21 +308,14 @@ void addRestClientBeans(Capabilities capabilities,
...
                         for (AnnotationInstance annotation : method.annotations()) {
                             if (annotation.target().kind() == AnnotationTarget.Kind.METHOD
                                     && !BUILTIN_HTTP_ANNOTATIONS_TO_METHOD.containsKey(annotation.name())
                                     && !ResteasyReactiveDotNames.PATH.equals(annotation.name())) {
-                                AnnotationValue value = annotation.value();
-                                if (value != null && value.kind() == AnnotationValue.Kind.ARRAY
-                                        && value.componentKind() == AnnotationValue.Kind.NESTED) {
-                                    for (AnnotationInstance annotationInstance : value.asNestedArray()) {
-                                        methodCreator.addAnnotation(annotationInstance);
-                                    }
```
Comment (Ladicek) [inline extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:321]: @michalszynkiewicz any reason why this was done? I can't think of anything.

**Comment by Ladicek** [comment]
Ah and one more thing: motivation for this is these 2 bugs @cescoffier found: https://github.com/smallrye/smallrye-fault-tolerance/issues/439 and https://github.com/smallrye/smallrye-fault-tolerance/issues/440

**Comment by geoand** [comment]
Nice!

I wonder if after this is in we can use it to replace things that seem similar in `org.jboss.resteasy.reactive.common.processor.AsmUtil`, `io.quarkus.deployment.util.AsmUtil` and `io.quarkus.arc.processor.AsmUtilCopy`

File: extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java (hunk 2 above)
Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:321]: I think simply doing `methodCreator.addAnnotation` didn't work for group annotations, e.g.:
```
    @ClientHeaderParam(name="OptionalMethodHeader", value="{fail}", required=false)
    @ClientHeaderParam(name="MethodHeaderExplicit", value="SomeValue")
    JsonObject methodOptionalMethodHeaderNotSentWhenComputeThrowsException();

```

Comment (michalszynkiewicz) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:321]: But that should be tested in the TCK, so if the TCK is passing without it either I was wrong or something got fixed in gizmo

Comment (Ladicek) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:321]: Oh I did fix annotation support in Gizmo for sure :-) See https://github.com/quarkusio/gizmo/pull/81

Comment (Ladicek) [reply extensions/resteasy-reactive/rest-client-reactive/deployment/src/main/java/io/quarkus/rest/client/reactive/deployment/RestClientReactiveProcessor.java:321]: I was just wondering how could this possibly have worked, as the previous code simply took container annotations and un-nested one level of them, which IMHO must have resulted in a really weird bytecode... but I guess we have something somewhere to recover from it :-)

**Comment by Ladicek** [comment]
> I wonder if after this is in we can use it to replace things that seem similar in `org.jboss.resteasy.reactive.common.processor.AsmUtil`, `io.quarkus.deployment.util.AsmUtil` and `io.quarkus.arc.processor.AsmUtilCopy`

Whoah! I didn't see those classes, as I only looked for usages of `SignatureElement.setSignature` (and missed the one usage in ArC).

It seems that the code in this PR is a bit more complete when it comes to method signatures (properly supports wildcards, inner classes, checked exceptions, and maybe something else), but it doesn't have the type variable substitution feature.

The very fact that we have 3 copies (4 with this PR) suggests that this belongs to Jandex. Eventually, we'll get there :-) In the meantime, I guess I will try to merge the code in this PR with the `AsmUtil` copy in `core-deployment`.

**Comment by geoand** [comment]
Awesome, thanks. 

I agree, your version does seem more complete than the others

**Comment by Ladicek** [comment]
Alright, merged the code into `AsmUtil`. Fortunately `org.jboss.resteasy.reactive.common.processor.AsmUtil` and `io.quarkus.deployment.util.AsmUtil` are identical, so I could keep them identical. `io.quarkus.arc.processor.AsmUtilCopy` has diverged a lot, so I didn't touch that.

File: core/deployment/src/main/java/io/quarkus/deployment/util/AsmUtil.java [hunk 3]
```diff
@@ -68,49 +72,168 @@
...
+            result.append(':');
+        }
+        for (Type bound : typeParameter.bounds()) {
+            result.append(':');
+            toSignature(result, bound, typeArgMapper, false);
+        }
+    }
+
+    private static boolean hasImplicitObjectBound(TypeVariable typeParameter) {
+        // TODO is there a better way? :-/
```
Comment (Ladicek) [inline core/deployment/src/main/java/io/quarkus/deployment/util/AsmUtil.java:225]: @n1hility any help here? This is the only thing I wasn't able to reconstruct from Jandex's public API.

**Comment by Ladicek** [comment]
CI passed in my fork, marking as ready for review.

**Review by stuartwdouglas — APPROVED** [review]

**Comment by stuartwdouglas** [comment]
Failure looks unrelated

**Comment by gsmet** [comment]
@geoand do you want to have a look at this one?
@Ladicek is this something that should be backported to 2.0?

**Comment by Ladicek** [comment]
Hmm yea, I guess if we had the RestClient Reactive fix in 2.0, that would be nice. Let me add the label, and see if we can make it. Thanks!

**Review by geoand — APPROVED** [review]

**Comment by gsmet** [comment]
@michalszynkiewicz confirmed me it looked good on Zulip. I'm going to merge it.

**Pull request merged**